### ISSUE TITLE
typescript error (enhancement needed)

### ISSUE THREAD

**Issue by huineng** [Izhaki/nodemon-webpack-plugin#141]
reference this file https://github.com/Izhaki/nodemon-webpack-plugin/blob/f8300fb71fac54cb642372ed9af481b4595d9a16/typings.d.ts

I'm getting typescript error

webpack 5.67
nodemon-webpack-plugin ^4.7.0

```
node_modules/nodemon-webpack-plugin/typings.d.ts:13:9 - error TS7006: Parameter 'compiler' implicitly has an 'any' type.

13   apply(compiler): void;
           ~~~~~~~~
```

either update to 
```
import { Settings as NodemonOptions } from 'nodemon';

export = NodemonPlugin;

declare class NodemonPlugin {
  constructor(options?: NodemonOptions);

  apply(compiler: any): void;
}
```

or 

```
import { Settings as NodemonOptions } from 'nodemon';
import { Compiler } from 'webpack';

export = NodemonPlugin;

declare class NodemonPlugin {
  constructor(options?: NodemonOptions);

  apply(compiler: Compiler): void;
}
```

thanks

**Comment by Izhaki** [comment]
Hmmmm.... this rings a bell. I think there were some issue with making this work with both webpack 4 and 5 (without having to split this plugin into two version). Regardless, your suggestion will work:

```
apply(compiler: any): void;
```

What is odd is that you get that error to begin with.

Your log clearly shows `node_modules/nodemon-webpack-plugin/` so why would you get a typecheck error on something in `node_modules` is beyond me. What does your `tsconfig.json` look like?

**Comment by huineng** [comment]
Thanks, let me clarify ..  the require('nodemon-webpack-plugin') of course still works and i'm still using this require util...

But long time ago i replaced all requires by imports to take full benefit from typescript.

and the below explains why i have a problem with the import. (so it's not the plugin that causes the problem, but the @types

```
There is a different issue, which is module resolution. 
By module resolution, I mean the compiler trying to understand what ns means in an import statement
like: import * ns from "mod". To do so, the compiler needs the definition of a module,
this could be a .ts file for your own code, or a .d.ts for an imported definition file.
If the file was found, it will be included regardless of whether it was excluded in the previous steps or not.
```

**Comment by Izhaki** [comment]
I still don't understand why you get the error you do.

The typescript config should exclude (and that is the default) `node_modules`.

This means that your _own code_ should be type-checked against the typings provided by external packages from node_modules. But the types of the node_modules themselves should not be type-checked (because each package may use a slightly different typescript config - your own `tsconfig` is likely not to be compatible with that).

Your error clearly shows that your setup is such that the some typechecking is done on node_modules. This hints that something is wrong with your configuration.

**Comment by huineng** [comment]
hi, thanks
But there's nothing wrong with my tsconfig
this is how typescript works especially when you start using import from instead of require()

there are several threads on this on this topic , maybe this one https://github.com/microsoft/TypeScript/issues/40426

now i can avoid this by setting "noImplicitAny" to false, but that was my initial question, as it is considered a best practice, just to add any to you code

thanks

**Comment by Izhaki** [comment]
Can you share a reproducible example?

I've created a [codesandbox](https://codesandbox.io/s/nodemon-plugin-typescript-test-4d2oo?file=/src/index.ts) and cannot reproduce.

PS. I'm not sure the issue you've linked is relevant, as the types are exported as `.d.ts` for this plugin. So, for all I know, typescript shouldn't typecheck it unless explicitly requested to do so.

Some minimal reproduction would really help to get to the bottom of this.

**Comment by huineng** [comment]
Thanks, so i added the following to your sandbox tsconfig (could not create a clone yet)

 "noImplicitAny": true, (codesandbox doesn't have this as default)
 "target": "esnext",

when you inspect the definition you will see the below. now when i build this with webpack, the build breaks.

![image](https://user-images.githubusercontent.com/3482445/152380176-9d3a0417-5fe6-4cb1-926e-f894e035edd2.png)

**Comment by Izhaki** [comment]
OK. This is great progress. Where do you inspect the definition? How do you get to it from the sandbox?

**Comment by Izhaki** [comment]
Wow. I guess you learn something new every day... How come you are the first to report this is beyond me...

I've created a minimal reproduction (no Webpack), with as single file:

```ts
// src/main.ts
import NodemonPlugin from "nodemon-webpack-plugin";

const plugin = new NodemonPlugin();
```

And the following `tsconfig.json`:

```json
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "moduleResolution": "node",
    "noImplicitAny": true,
    "target": "esnext"
  }
}
```

Then run 

```shell
tsc --noEmit
```

And it complains.

Neither the "include" nor "exclude" options in `tsconfig` make a difference. But you can disable it with the following compiler option:

```json
    "skipLibCheck": true
```

Fixing this now.

**Comment by Izhaki** [comment]
Landed as 4.7.1 (and tested locally).

@huineng Thank you so much for reporting this and spending the time on this discussion!
